### DISCUSSION TITLE
Possible to generate one output file (entry) and still use split chunks?

### DISCUSSION THREAD

**Discussion by noahnu** [webpack/webpack#14421] [Q&A] [answered]
I have a single entrypoint with split chunks enabled. I want to generate a single output file "app-[contenthash].js" which will then import the other chunks as necessary. I've tried tweaking the config, and so far the only thing that seems to produce the single file is to disable split chunks entirely. With my current config, I'll get more than one `app-<some hash>.js` file.

My use case: I need a URI to a single script that I can dynamically inject into a page. Related to module federation.

---

## Key parts from my config

```
{
    entry: { app: ['entry.js] },
    output: {
        path: 'dist',
        publicPath: '/',
        filename: '[name]-[contenthash].js',
        chunkFilename: 'chunk-[contenthash].js',
        sourceMapFilename: '[contenthash].js.map',
    },
    optimization: {
        splitChunks: {
            name: false,
            minChunks: 2,
            chunks: 'async',
            maxSize: 10000000,
            minSize: 20000,
            maxInitialRequests: 30,
            maxAsyncRequests: 30,
            automaticNameDelimiter: '-',
            cacheGroups: {
                defaultVendors: {
                    priority: 0,
                    name: 'vendors',
                    test: /[\\/]node_modules[\\/]/i,
                },
            },
        },
    },
}
```

## Related

This seems to be what I want, but the question wasn't given a satisfactory answer: https://stackoverflow.com/questions/53187492/webpack-how-to-create-split-bundle-with-a-single-entrypoint

**Comment by alexander-akait** [comment, marked as answer]
What is sense to use `splitChunks` and union them to one file?

**Reply by noahnu** [reply]
Just the entry point. So we'd generate a single "app-[contenthash].js" file which can then lazy load the rest of the split chunks. If we disable split chunks, then we end up having to download a very large file over the network, which means it takes longer for the code to start executing and the app's performance suffers. Some of the lazy loaded chunks might not even be necessary depending on some environment conditions, so they'll never be downloaded. Same as a regular split chunk case.

**Reply by alexander-akait** [reply]
But why you use ` chunks: 'all',`?

**Reply by noahnu** [reply]
So to be clear, I don't want to union them to one file. I want them to be imported from that file.

**Reply by noahnu** [reply]
Oh, this is partly from a different config I use. I did try `chunks: 'async'`, but it doesn't seem to make a difference. I still get 2 "app" files. **edit**: updated config to show "async"

**Reply by alexander-akait** [reply]
If you will not touch initial chunks they will be already in your bundle and you don't need to load them

**Reply by alexander-akait** [reply]
Because async should be loaded asynchronously, so you have them and them load them when it will be needed

**Reply by alexander-akait** [reply]
Set `output.publicPath: 'auto'` and keep two file, webpack load extra file when it will be needed, so you will have small initial chunk and load async chunks when it will be needed

**Reply by noahnu** [reply]
Not sure I understand. It's still creating 2 initial chunks?

Would another way to work around this issue be to change the entry to just a single line like this? So the initial chunk becomes minimal and everything else gets marked as async?

```
import('entry.js')
```

**Reply by alexander-akait** [reply]
Disable default https://github.com/webpack/webpack/blob/main/lib/config/defaults.js#L1070, i.e. `false`, yes you can conver them to `import("entry.js")`, but it can be impossible for some cases

**Reply by noahnu** [reply]
The follow up question would be whether I can disable split chunks for the initial chunk and only split async? Even when I set `chunks: 'async'` it's still splitting the initial "app", or at least naming the other chunk as "app" so I can't tell them apart.

**Reply by noahnu** [reply]
Setting `default: false` and `chunks: 'async'` still produces:

![image](https://user-images.githubusercontent.com/1297096/136256367-207a3e73-176b-437c-a044-7fc35cd5060f.png)

**Reply by alexander-akait** [reply]
> The follow up question would be whether I can disable split chunks for the initial chunk and only split async? Even when I set chunks: 'async' it's still splitting the initial "app", or at least naming the other chunk as "app" so I can't tell them apart.

Yes, just try different options

> Setting default: false and chunks: 'async' still produces:

Just disable this warning, expected, you have big initial chunk, you need rewrite it on more `import()` to reduce size

**Reply by noahnu** [reply]
> Just disable this warning

I was referring to the 2 "app-<hash>.js" files rather than the warning.

Either way, the `import("entry.js")` approach seems to work well.

**Reply by alexander-akait** [reply]
hm, I see you extract css, we have bug for extracting - extra js file, I think you are faced with this, hope we will fix it in near future

**Reply by noahnu** [reply]
Do you recall if there's a github issue for that? couldn't find anything with a quick search

**Reply by alexander-akait** [reply]
https://github.com/webpack/webpack/issues/11671

**Reply by alexander-akait** [reply]
If you provide example of the problem I will investigate

**Reply by noahnu** [reply]
I'll try put together a reproduction. My particular project is closed source, but I can copy some of the config.
